fix(releases): Allow project:releases scope for org releases endpoint

szokeasaurusrex opened the pull request:
## Changes

- Add authorization logic to allow users with `project:releases` or `org:ci` scopes to create releases via the organization releases endpoint, even if they're not direct team members of the project
- Ensures consistency with the project releases endpoint behavior
- Add test case to verify users with `project:releases` token can create releases for projects they're not team members of
- Fix test to use proper silo mode context (`assume_test_silo_mode(SiloMode.CONTROL)`) for `ApiToken` creation

## Context

Previously, the organization releases endpoint only allowed release creation if the user was either a member of the project or was using an org Auth Token with `org:ci` scope. This change extends that check to also include users with `project:releases` scope, matching the behavior of the project releases endpoint and providing consistency across the API.

## Issues

- Fixes [BE-611](https://linear.app/getsentry/issue/BE-611/fix-error-with-user-auth-tokens)
- Ref #105038

szokeasaurusrex committed:
ref(releases): Allow project:releases scope for org releases endpoint

- Add logic to check for project:releases and org:ci scopes when determining release permissions
- Ensures consistency with project releases endpoint behavior
- Add test to verify users with project:releases token can create releases for non-member projects
- Fix test to use proper silo mode context for ApiToken creation

szokeasaurusrex committed:
fix

szokeasaurusrex committed:
TypeGuard to fix mypy

loewenheim commented on src/sentry/api/bases/organization.py:
```diff
@@ -669,6 +669,14 @@ def get_projects(  # type: ignore[override]
+        if not has_valid_api_key and request.user and request.user.is_authenticated:
+            if request.access.has_scope("project:releases") or request.access.has_scope("org:ci"):
+                has_valid_api_key = True
+
```
This seems to just reiterate the comment above the change, is that correct @szokeasaurusrex ?

loewenheim commented on tests/sentry/api/endpoints/test_organization_releases.py:
```diff
@@ -2101,6 +2101,61 @@ def test_bad_repo_name(self) -> None:
+        team1 = self.create_team(organization=org)
+        team2 = self.create_team(organization=org)
+
+        # User is only a member of team1
```
This comment is confusing because at this point the user hasn't been added to any team yet.
```suggestion
```

szokeasaurusrex committed:
Apply suggestion from @loewenheim

Co-authored-by: Sebastian Zivota <loewenheim@users.noreply.github.com>

szokeasaurusrex committed:
fix: Use request.access

request.access checks organization membership, and only returns scopes if the token should be able to acccess the given organization.

szokeasaurusrex committed:
Add test for proper use of `request.access`

szokeasaurusrex commented on src/sentry/api/bases/organization.py:
```diff
@@ -669,6 +670,14 @@ def get_projects(  # type: ignore[override]
+        # is not a direct team member of the project.
+        has_valid_api_key = has_valid_api_key or (
+            is_api_token_auth(request.auth) and request.auth.has_scope("project:releases")
+        )
```
All current code paths should be preventing cross-org access. However, we should probably validate this permission also within the function itself to be sure that any future codepaths calling this function don't introduce cross-organization access.

I have solved the issue by using `request.access`, which factors organization membership into the scopes (when the user is not a member of the given org, `request.access` returns an empty list of auth scopes). I also added a test which ensures cross-organization access is impossible.

szokeasaurusrex commented on src/sentry/api/bases/organization.py:
```diff
@@ -669,6 +670,14 @@ def get_projects(  # type: ignore[override]
+        # is not a direct team member of the project.
+        has_valid_api_key = has_valid_api_key or (
+            is_api_token_auth(request.auth) and request.access.has_scope("project:releases")
+        )
```
This is wrong. The tests validate that the logic works as expected.
